hail/src/main/scala/is/hail/backend/Backend.scala

```diff
@@ -94,5 +96,36 @@ abstract class Backend {
     Serialization.write(Map("value" -> jsonValue, "timings" -> timings.value))(new DefaultFormats {})
   }
 
+  def encode(ir0: IR, codecString: String): (String, Array[Byte]) = {
+    val codec = CodecSpec.fromString(codecString)
+    val ir = lower(ir0, None, false)
+
+    if (!Compilable(ir))
+      throw new LowererUnsupportedOperation(s"lowered to uncompilable IR: ${Pretty(ir)}")
+    if (ir.typ == TVoid)
+      throw new LowererUnsupportedOperation(s"lowered to unit-valued IR: ${Pretty(ir)}")
+
+    Region.scoped { region =>
+      val (pt: PTuple, f) = Compile[Long](MakeTuple(FastSeq(ir)))
+      (pt.parsableString(), codec.encode(pt, region, f(0)(region)))
```

Comment:
What's the index 0 for? I tried to tell from the code.

Reply:
partition index, used for randomness